src/registry/otlp/attribute_otlp_ingester.rs

```diff
@@ -0,0 +1,193 @@
+use std::time::Duration;
+
+use serde_json::{json, Value};
+use weaver_common::Logger;
+use weaver_live_check::{sample::SampleAttribute, Error, Ingester};
+
+use super::{grpc_stubs::proto::common::v1::AnyValue, listen_otlp_requests, OtlpRequest};
+
+/// An ingester for OTLP data
+pub struct AttributeOtlpIngester {
+    /// The address of the OTLP gRPC server
+    pub otlp_grpc_address: String,
+    /// The port of the OTLP gRPC server
+    pub otlp_grpc_port: u16,
+    /// The port of the admin server
+    pub admin_port: u16,
+    /// The inactivity timeout
+    pub inactivity_timeout: u64,
+}
+
+impl AttributeOtlpIngester {
+    fn maybe_to_json(value: Option<AnyValue>) -> Option<Value> {
+        if let Some(value) = value {
+            if let Some(value) = value.value {
+                use crate::registry::otlp::grpc_stubs::proto::common::v1::any_value::Value as GrpcValue;
+                match value {
+                    GrpcValue::StringValue(string) => Some(Value::String(string)),
+                    GrpcValue::IntValue(int_value) => Some(Value::Number(int_value.into())),
+                    GrpcValue::DoubleValue(double_value) => Some(json!(double_value)),
+                    GrpcValue::BoolValue(bool_value) => Some(Value::Bool(bool_value)),
+                    GrpcValue::ArrayValue(array_value) => {
+                        let mut vec = Vec::new();
+                        for value in array_value.values {
+                            if let Some(value) = Self::maybe_to_json(Some(value)) {
+                                vec.push(value);
+                            }
+                        }
+                        Some(Value::Array(vec))
+                    }
+                    _ => None,
+                }
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+}
+
+/// Iterator for OTLP attributes
+struct OtlpAttributeIterator {
+    otlp_requests: Box<dyn Iterator<Item = OtlpRequest>>,
+    buffer: Vec<SampleAttribute>,
+}
+
+impl OtlpAttributeIterator {
+    fn new(otlp_requests: Box<dyn Iterator<Item = OtlpRequest>>) -> Self {
+        Self {
+            otlp_requests,
+            buffer: Vec::new(),
+        }
+    }
+
+    fn extract_attributes_from_request(&mut self, request: OtlpRequest) -> Option<SampleAttribute> {
```

Comment:
Nit: I'd find this class less confusing if it was shifted a bit:

- Next very clearly attempts to have a buffer *before* returning next.
- Methods like this are responsible for filling the buffer.

```
 /// Puts sample attributes in self.buffer from a request.
 fn fill_attribute_buffer_from_requests(&mut self, request: OtlpRequest) {
   ...
 }
 ... impl Iterator ...
   fn next(&mut self) -> Option<Self::Item> {
     while (self.buffer.empty()) {
        match self.otlp_requests.next() {
            Some(request) => self.extract_attributes_from_request(request),
            None => return None,
        }
     }
     // We know we have a buffer at this point.
     Some(self.buffer.remove(0));
   }
```